### DISCUSSION TITLE
DRC errors in detailed routing - "Metal spacing violation"

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3216] [General]
I think this is happening because two macros that need straight vertical wires between them don't have IO pins lined up and then the wires between the macros are criss-crossing.

What is the difference and purpose of `MACRO_PLACE_HALO` vs `MACRO_PLACE_CHANNEL` and where do I use them, in the macro or where I'm using the macro from?

Some thoughts on what I might do:

- I could modify the halo macro to have a big enough "halo" (black area) where the wires can move horizontally to line up better with their destination pin in the other macro.
- I could fix the io pin locations and floorplan so the router is able to use straight vertical wires instead of crisscrossing between two macros. This would reduce the need for a big halo.

![image](https://user-images.githubusercontent.com/2798822/233312554-d99551bf-439f-47f9-85da-21b601517284.png)

**Comment by vijayank88** [comment]
@oharboe 
Variable description found here: https://openroad-flow-scripts.readthedocs.io/en/latest/user/FlowVariables.html#floorplan

**Reply by oharboe** [reply]
Yes... but I'm wondering what it means?

I can increase the "halo", but what is a halo and what effect does it have? Do I specify this on the macro in the config.mk where I instantiate the macro.

Same for MACRO_PLACE_CHANNEL.

MACRO_PLACE_HALO | horizontal/vertical halo around macros (microns).
-- | --
MACRO_PLACE_CHANNEL | horizontal/vertical channel width between macros (microns).

**Reply by vijayank88** [reply]
It is useful where automatic macro placement technique used. 
If you're using manual macro placement above variable not in effective.
HALO/CHANNEL provide enough space for routing and where std cells not placed.

**Reply by oharboe** [reply]
I see. I am still not clear on the distinction between channel and halo though.

**Comment by maliberty** [comment]
Note that mpl distinguished halo vs channel but mpl2 just uses halo.  The distinction is that a channel is the space between two macros but the halo applies on all sides of the macro.

I can't tell much from the picture.

**Reply by oharboe** [reply]
From your explanation, I would have thought that you specify *either* HALO or CHANNEL and that HALO took just one integer?

```
export MACRO_PLACE_HALO = 10 10
export MACRO_PLACE_CHANNEL = 20 20
```

**Comment by maliberty** [comment]
The setting are for x & y separately.   Usually they are the same

**Reply by oharboe** [reply]
Then I'm still confused about the difference between MACRO_PLACE_CHANNEL and MACRO_PLACE_HALO.

**Comment by maliberty** [comment]
Imagine channel=10 and halo=5.  Then macros must be 10 apart but standard cells must be 5 away from a macro.  It is moot with mpl2.

**Comment by oharboe** [comment]
Updated docs in pull request based on the above: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1007
